### DISCUSSION TITLE
Optional rule suggestion: ban implicit truthy/falsey type conversion in truth value testing

### DISCUSSION THREAD

**Discussion by kkom** [microsoft/pyright#6488] [Ideas]
I've started [a discussion on this](https://github.com/astral-sh/ruff/discussions/8757) on Ruff's GitHub, but I'm also curious if this is something that Pyright might consider implementing.

I'd like to ban implicit truthy/falsy checks in our codebase, to avoid common bugs like this:

```python
def maybe_int() -> int | None:
    """returns None if data is missing, otherwise the value"""
    return 0

x = maybe_int()

if not x:
    return "early return because we're missing data"

# we now have a bug, because `0` is a perfectly fine value, but it was discarded in the early return above

process_data(x)
```

More details and broader rationale in this discussion: https://github.com/astral-sh/ruff/discussions/8757

@erictraut – what's you view on this? Does this feel like a rule Pyright might implement? (I would imagine it would be optional even in strict mode.)

On one hand it feels opinionated enough to be a linter rule, on the other hand I have a hunch it might be easier and more robust to implement in a type checker (feels a bit similar to e.g. `reportUnnecessaryComparison`).

**Comment by erictraut** [comment]
This pattern has legitimate uses and is quite common in Python, so it's not a rule that could be enabled by default, even in strict mode. In general, I don't like adding rules that cannot be enabled by default in strict mode because they are difficult for pyright users to discover and therefore are rarely used. This situation is also not a type violation, which means that it's not obvious that it's a rule a type checker should be enforcing.

There is a precedent for pyright supporting rules that are not type violations but do require type analysis (and are therefore difficult for linters to implement). These rules all involve cases where 1) there's a relatively low chance of a false positive and 2) a reasonably high likelihood of an actual bug. The `reportUnnecessaryComparison` rule is a good example of such a rule.

This particular check doesn't seem to meet this bar. I'm not convinced that it represents a sufficiently common source of bugs to merit adding such a rule, but I could perhaps be persuaded with some additional evidence.

For comparison, the typescript compiler (which is effectively a static type checker) does not provide an option like the one you're proposing, but eslint (which is more akin to a ruff) [does](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md).

**Reply by kkom** [reply]
Thank you for the response. I understand the discoverability problem of beyond-strict rules...

When I received your response, I was typing an additional comment giving an example why I think the type-checker is better suited for a rule like this. Consider code like this:

```python
def maybe_bool() -> bool | None:
    ...

x = maybe_bool()

if x is None:
    return "early return because we're missing information"

if x:
    # do something because `x` is `True`
else:
    # do something because `x` is `False`
```

I don't see like a linter could understand that the `if x:` line is fine (because the type has been narrowed to `bool` from `bool | None`). I assume that such type narrowing is the domain of the typechecker, rather than linter.

I have a follow-on thought. Have you seen close integration between a type-checker and a linter in the past? The type checker could provide type inference functionality, which the linter would use to implemented opinionated rules like the one I proposed.

If this hasn't been done before, what are you thoughts on this?

(I'm a happy user of both Ruff and Pyright, and while I'm not involved with any of the projects directly, I think an integration like this could unlock a lot of benefits.)

**Reply by kkom** [reply]
In terms of any additional validation, I can share that Facebook has enabled a rule like this in its PHP/Hack codebase (through the HHVM typechecker) a few years ago. It's a little confusingly called `sketchy null check`, but has worked well in the codebase.

See these release notes for some reference: https://hhvm.com/blog/2018/08/28/hhvm-3.28.0.html

**Reply by erictraut** [reply]
> Have you seen close integration between a type-checker and a linter in the past? The type checker could provide type inference functionality, which the linter would use to implemented opinionated rules like the one I proposed.

It's an interesting idea. In theory, ruff could do that today using pyright's language server interface. It would do this by instantiating the pyright language server and then passing it a "hover text" request for a particular position within the file. Pyright would then evaluate the type of the identifier at that location and pass back the hover text, which includes the identifier's narrowed type. The downside is that the output of the hover text LSP command is _text_ (specifically, _markdown_), so ruff would need to parse the text and convert it back into its own representation of a type.

In theory, an alternative interface could be added to pyright that effectively does the same thing as the "hover text" request but returns type information in a manner that doesn't require conversion from text. The encoding would need to be agreed upon by tools vendors — and preferably standardized in a specification. There's currently no provision in the language server protocol for encoding type information. That's understandable given that the LSP is language-agnostic, but types are language-specific.

If you or someone else is interested in prototyping this idea using the current facilities in pyright, you could look at my implementation of the web-based [pyright playground](https://pyright-play.net/). It does something similar to what I'm describing above. It installs the pyright language server from the npm registry and [instantiates it](https://github.com/erictraut/pyright-playground/blob/015ad4933fb5c6fe18ed403958fde3ea547b92e2/server/src/sessionManager.ts#L124), then sends it LSP messages to retrieve diagnostics, hover text, completion suggestions, etc.

**Reply by hmc-cs-mdrissi** [reply]
One idea I've considered for building on top of pyright to use type of expressions for custom logic/tooling was to take code like this,

```python
def foo(x: int) -> str:
  y = "a"
  z = y + str(x)
  return z
```

and then rewrite ast (ast.parse -> NodeTransformer -> ast.unparse if done in python) and dump it into something like,

```python
def foo(x: int) -> str:
  y = reveal_type("a")
  z = reveal_type(y + str(x))
  return z
```

Exactly how much to reveal_type depends on which nodes you need type information for (assignments, if condition, return expresssion, etc). Then run pyright --outputjson on the new file and extract the type information from json.

For this specific idea it should be possible to parse AST -> add reveal_type for every if statements condition -> run pyright -> extract type information for if statements -> define lint rule. This is roughly taking python AST and creating an AST + type information and more reveal_types you put in the closer it becomes to that.

**Reply by patrick-kidger** [reply]
> but I could perhaps be persuaded with some additional evidence.

Since this was mentioned, I'd like to chime in with an additional +1 for this feature.

The examples given elsewhere in this thread are exactly the ones that I find people bumping into. Here's my own flavour of this:
```python
def f(random_seed: Optional[int]):
    if not random_seed:
        random_seed = seed_from_system_time()
    ...    # do stuff deterministically on `random_seed`

f(0)  # Silently not deterministic!
```

**Reply by npip99** [reply]
I think it's odd that pyright won't implement any flags that don't meet this bar:
1) there's a relatively low chance of a false positive
2) a reasonably high likelihood of an actual bug

I want heavy type-checking to the level of a static typechecker, not because dynamic typing creates bugs (which it does), but because dynamic typing is impossible to maintain and very difficult to read.

Knowing that `if x:` is short-hand for `if (isinstance(x, str) and len(x) > 0) or (isinstance(x, list) and len(x) > 0) or x is not None:` is very difficult to read and debug. It's exhausting to be reading a segment of code and constantly be thinking about empty lists and empty strings in the back of your head when understanding what the function does. I'd rather just have it be written out explicitly, usually, `if x is not None:` or `if len(x) > 0:` is what is actually intended, and it becomes very easy to read and understand.

"Low chance of false positive + high likelihood of an actual bug", sounds like the bar that would be given to the weakest possible flag available. Not the strongest and strictest flag. Why wouldn't `no-implicit-boolean-coercion` be viable for the strongest possible flag?

**Reply by erictraut** [reply]
@npip99, the general rule that I use for determining whether to implement a check includes the two requirements you've listed plus the following:

3. It's not possible to implement using a simple linter rule. Pyright is not a linter, it's a type checker.
4. It is a rule that could be enabled by default in strict mode. If it can't be enabled by default in strict mode (because it's too esoteric or dictates a nonstandard coding approach), it will rarely be used because of discoverability issues. 

You mentioned that you would prefer to write out `isinstance(x, str) and len(x) > 0`. I think it's safe to say that most Python developers would not want this forced upon them if they use strict-mode type checking.

Decisions like this often come down to a judgement call on my part. I'm willing to be proven wrong with my decisions based on feedback from pyright users, but I've been doing this long enough that my initial instincts have proven to be correct most of the time.

**Reply by npip99** [reply]
@erictraut It's not that I'd prefer to write out `isinstance(x, str) and len(x) > 0`, it's that `if x:` implicitly writes that, and I never want it to write that. I just want to check `if x is not None:`. It's hard for me to say what other devs want, but for me, about half the time I see `if x:` in a codebase, the purpose of that line of code is to check if something is not None. And, in all of those "None" checks, `if x:`'s implicit ability to check `(isinstance(x, list) and len(x) > 0) or x == 0.0` is a bug 100% of the time and has no false positives. The resulting "len(x) > 0 or x == 0" bug is just often benign for no other reason other than that empty strings and the number 0 are rare (Similar to SQL bugs that arise when someone's last name is literally `null` - it's just unlikely).

Implicit boolean is also a bad practice that's recognized by [PEP8](https://peps.python.org/pep-0008/#programming-recommendations), but currently neither pyright nor ruff can lint my codebases to ensure that they're adhering to that recommendation.

I'm also a developer that very happily uses `reportImplicitStringConcatenation`, `reportImplicitOverride `, `reportMissingSuperCall`, `reportShadowedImports`, `reportUninitializedInstanceVariable`, none of which are set by default when using `strict` mode. So, I really think a `reportImplicitBoolean` would be very fitting to join those others. I think it's actually more important than any of the others in that list (reports for string concat / implicit override / missing super are all nice, but less important than implicit boolean imo).

**Reply by kkom** [reply]
@erictraut – please forgive me for replying to a discussion I originally started, but IMO @npip99 brings up an excellent point I wasn't aware of. Let me copy the relevant part of [PEP 8](https://peps.python.org/pep-0008/#programming-recommendations) for posterity:

> Comparisons to singletons like None should always be done with `is` or `is not`, never the equality operators.
>
> Also, beware of writing `if x` when you really mean `if x is not None` – e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

In addition to supporting @npip99's assessment that many `if x:` statements when `x` isn't a bool are PEP8 violations, because the programmer really means `if x is not None:` (in my experience it's even much more than 50%), I'd add that such a setting would be synergistic with `reportUnnecessaryComparison`.

I've seen many cases where upgrading from a simple `if x:` to `if x is not None:` highlights that entire conditional is dead code – because the type of `x` was narrowed to statically exclude `None` some time after writing the original comparison.

_Edit:_ This also means that for users of Pyright `if x:` in practice would be very rarely replaced by the verbose forms of `if x is not None and len(x) > 0:` or more formally `if isinstance(x, Sized) and len(x) > 0:`. Pyright would usually flag that first condition as unnecessary. I hope this alleviates the concern you raised.

PS: If such a rule were to be added, objects implementing the [`__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method should naturally be excluded from it.

**Reply by kkom** [reply]
Another common case is doing `if x:` when `x` is a not-None collection. In such a case writing `if len(x) > 0:` is also easier to read and less error-prone.

Somewhat analogously to the `reportUnnecessaryComparison` point above, explicit `if len(x) > 0:` has the advantage that the type checker highlights when it becomes possible that `x` could be None. This allows the programmer to properly consider that scenario, which might be to e.g. load a value missing in memory – rather than simply assume that the check should fail.

FWIW,  I realise we're entering a bit of [PEP 661](https://peps.python.org/pep-0661/) territory here. However, I wouldn't speculate about the outcome of that proposal. Today in Python `None` is commonly used as a special value/type to indicate missing data.

**Reply by npip99** [reply]
Today I just did a full audit of a codebase I maintain to check the impact of this rule:

---

```
(backend) npip99@npip99:~/Documents/programming/...$ find . -name '*.py' | xargs wc -l | tail -n1
   15418 total
(backend) npip99@npip99:~/Documents/programming/...$ find . -name '*.py' | xargs grep -E '^\s*if.*:$' | wc -l
   596
```

15k LOC, with 596 if statements.

---

With regexes and manual inspection, there were 340 irrelevant if statements (random boolean algebra), and 256 relevant if statements (None check, empty string, list non-empty). Of the 256 relevant if statements, here was the breakdown:

<img width="75%" src="https://github.com/user-attachments/assets/302676cc-68e3-4c18-b57a-573855b85f5d" />

@kkom was right, `None` checks are indeed far more than 50%. They're actually 82.3% of all relevant if statements in my sample. I already invest a lot of mental effort to review every PR and check every if statement for None checks, and despite this, 7% of all None checks were PEP8 violations. *Checking for PEP8 violations is crucial, 20% of all PEP8 violations were bugs.* Here is the offending code:

<img width="75%" src="https://github.com/user-attachments/assets/b2b13c51-faf5-449e-8dd5-98475bf3faa2" />

The fact that empty strings are falsy, creates a bug in this situation. A benign bug, because empty strings are rarely useful. This code will end up throwing an HTTP 500 error in the empty string case, which still is not ideal when trying to build robust backend applications.
- The fact that empty strings are falsy is frustrating for me because it really has absolutely no benefit - out of 15,000 LOC, I only intentionally check for an empty string once. So, the ability to do `if x:` instead of `if x != "":` only saves 4 characters per 15k LOC, and even for that 1 line of code it's a negative impact because how it harms readability. Imo the idea that empty strings are falsy is always bad - an empty string check should always be done with an explicit `!= ""`, and if a line of code uses the fact that empty strings are falsy, then it has a virtually 100% chance of being a bug.

Arrays are more debatable; 16.4% of relevant if statements were `len` checks. Personally, I do `if len(x) > 0:`, because it's more readable. `if x:` would save 7 characters, or 294 characters total across the codebase. I don't consider the loss in readability to be worth it, and using the `if x:` shorthand would definitely be considered "not strict" programming imo.

- Also, if you accidentally write `if x:` where x is an Optional[list], then nothing will complain, even though it has a virtually 100% chance of being a bug. So, `if len(x) > 0:` is the only way to get the typechecker to correctly complain if len-checking an Optional variable.

---

Addendum:

> PS: If such a rule were to be added, objects implementing the [__bool__](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method should naturally be excluded from it.

This never occurred in my codebase, I'm not sure in what situation this would happen in. `if bool(x):` seems fine for those rare situations, it's always good practice to be explicit in rare situations imo. (For example, even the always-truthy `__bool__` of a [regex's match](https://docs.python.org/3/library/re.html#re.match) object, is still handled better with `if x is not None:`.)

**Reply by marcuslimdw** [reply]
My two cents: implicit boolean conversions have caused at least two bugs at my workplace that I know of in the last 6 months, and I would 100% use this rule.

**Reply by erictraut** [reply]
> PS: If such a rule were to be added, objects implementing the [bool](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method should naturally be excluded from it.

I'll note that the `int` and `float` classes implement the `__bool__` method, so if such a check were to exempt classes that implement the `__bool__` method, it wouldn't apply to some of the cases mentioned above. It would still apply to `str`, `list`, `tuple` and other instances of `Sequence` because they implement `__len__` rather than `__bool__`.

**Reply by kkom** [reply]
Ah, thanks for pointing it out @erictraut!

~That's unfortunate (because I think that practically implicit bool conversions for numbers are undesirable as well), but I think still better do what you say for the sake of consistency / soundness.~

~Seeing the rule fire in most situations would still build some awareness of dangerous conditionals in a codebase or team.~

~**Edit:** `if x:` would still be captured in the most common undesirable case where `x: int | None`, because `None` doesn't implement `__bool__`, right?~

~**Edit 2:** However, if you call would be to go even stricter, and ban even relying on an existing `__bool__` method, I personally wouldn't be bothered by it! I agree with @npip99 that wrapping things with an explicit `bool(...)` would lead to clearer code in these scenarios.~

I've changed my position here after a bit of research. It turns out that `None` also implements `__bool__()`:

```python
>>> None.__bool__()
False
>>> object().__bool__()
Traceback (most recent call last):
  File "<python-input-5>", line 1, in <module>
    object().__bool__()
    ^^^^^^^^^^^^^^^^^
AttributeError: 'object' object has no attribute '__bool__'. Did you mean: '__doc__'?
```

This means that neither of the examples below:

```python
x: int

if x:
    ...

y: int | None

if y:
    ...
```

would be disallowed by the rule we're discussing here. And the second is in my experience a very common source of bugs.

I think then practically it would be desirable to lean towards strictness, and ban both the `if __len__() > 0` and `if __bool__()` truthiness heuristics. On second though, that feels consistent as well – we're disregarding all heuristics, and require only explicit type matches.

As @npip99 pointed out, there is always the option to wrap a statement in `bool(...)` if someone desires type coercion to a boolean.

**Comment by nurpax** [comment]
I have a case where an API some functions return a direct `bool` and some functions `tuple[bool, bool]`.  A user might write:

```
if some_func():  # return value of some_func is tuple[bool, bool]
    # ...
```

when they actually meant:

```
if some_func()[0]:
    # ...
```

It feels unfortunate that pyright doesn't seem to have an option to warn about "truthy tuples", esp. as the type of this tuple is statically determined and known to always evaluate to True.

Ruff has a rule for this (https://docs.astral.sh/ruff/rules/if-tuple/) but unfortunately it only works for literal values, as Ruff is not a type checker.

**Comment by shartzog** [comment]
I'd support an optional rule to flag this sort of usage, but I'd like to reiterate that a 'fast and loose association with the truth' has been a part of python since its inception (for good or ill), and it *can* be a perfectly legitimate shorthand. E.g. regex usage:
```python
animal_pattern = re.compile(r"(?P<phrase>(?P<prefix>It's|This is an? )(?P<animal>\w+)!)")
animals: list[dict[str, str]] = []

phrase = "This is a walrus!"

if _animal_match := animal_pattern.match(phrase):  # seems legit...
    animals.append(_animal_match.groupdict())

if (_animal_match := animal_pattern.match(phrase)) is not None:  # feels clunky and unnecessary...
    animals.append(_animal_match.groupdict())
```
Strict container truthiness might also encourage the introduction of bugs from simple type system frustration. For example, a frustrating or cryptic flag might lead to bad decisions like omitting the `my_list and` from a conditional such as:
```python
my_list: list[int]
if my_list and my_list[0] < 10:
    my_list[0] += 10
```
You'd want to be *very* explicit about _why_ this rule exists and how to fix it in the error message.

**Comment by yamplum** [comment]
A lack of such a check could lead to silent introduction of bugs during refactoring.

Suppose I have the following code:
```python
def try_frobnicate() -> bool:
    # ...

# elsewhere
success = try_frobnicate()

if success: # correct usage - success is a bool
    # ...
```

Then later I refactor it to provide more information:
```python
def try_frobnicate() -> Result[Success, Error]: # Result is Union[Ok, Err]
    # ...

# elsewhere
success = try_frobnicate()

if success: # uh-oh, Err is truthy but no warning raised, hard to spot manually
    # ...
```

I think any codebase that takes type safety seriously will (or at least should) want to ban non-bool values from condition checks. I'm not sure if it is feasible at a linter level.

**Comment by kkom** [comment]
@erictraut – I remember you saying that you could revisit decisions to implement new rules if there is evidence of community demand (upvotes, comments). Could I ask what's your view on this rule now?

It's among the [top open discussions](https://github.com/microsoft/pyright/discussions/categories/ideas?discussions_q=is%3Aopen+sort%3Atop+category%3AIdeas) at the moment. Especially if, in addition to "upvotes", you also count "thumbs up" – which this discussion seems to have more relatively to others.

**Edit:** I think a succinct, Pythonic justification for this rule could be that it prevents PEP 8 violations, which [states](https://peps.python.org/pep-0008/#programming-recommendations):

> beware of writing `if x` when you really mean `if x is not None` – e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

**Comment by Timmmm** [comment]
Just ran into a bug that this would have caught. I was refactoring some code and they had used an implicit bool coersion like this:

```
id = get_value_from_file()
if id:
    ...
```

During the refactoring I change the type of `id` from `string | None` to `int`. That caused no type errors at this bit of code, but it silently broke the case when `id` is 0.

I'm surprised this isn't a rule already. It's a pretty common footgun in my experience, and very bad style so nobody should be writing code like this anyway (so IMO it should be enabled by default in strict mode). It's also difficult for tools like Pylint to implement because it requires precise type knowledge.

Typescript has had [this rule available for a long time](https://typescript-eslint.io/rules/strict-boolean-expressions/).

**Reply by shartzog** [reply]
The problem in your case isn't necessarily the implicit `__bool__` call. Would you still expect a typing error if your original code had been written as follows?
```python
id = get_value_from_file()
if bool(id):
    ...
```
Long story short, this rule would more or less invalidate Python's `__bool__` methodology entirely. Long time python devs are very aware of this pitfall and approach str --> int conversions with care, and many if not most consider the implicit case you've raised to be the *preferred* styling. IMHO, the biggest issue with your sample code is that it masks the builtin `id` function, not the implicit `__bool__` call. 🤷

In any case, I definitely support an *optional* rule as stated earlier in this thread.

**Reply by Timmmm** [reply]
> The problem in your case isn't necessarily the implicit __bool__ call.

Yes it is. If implicit bools were not allowed then the author would likely have written `if id is not None and id != 0:`.

> `if bool(id):`

I would not expect a typing error in that case, but I have also never seen anyone write this, whereas implicit bool coersion from `str`, `int`, `None` and `list` is an *extremely* common mistake.

> Long time python devs are very aware of this pitfall and approach str --> int conversions with care

Indeed. I am not working with long-time Python devs. Also that is very much a "I don't need seatbelts; I'm a good driver" attitude.

> many if not most consider the implicit case you've raised to be the preferred styling

Then they are misguided. Many (if not most) Python devs consider static typing to be a waste of time entirely.

> it masks the builtin `id` function

It was just an example. The actual code doesn't use `id`.

**Reply by shartzog** [reply]
> I would not expect a typing error in that case, but I have also never seen anyone write this

I've come across a fair bit of `bool(str)` and `bool(list)`, but not in simple conditionals to your point. It's usually for type narrowing a function input or output when a type checker complains and usually in a functional(ish) programming context. Not saying it's "right" but cheap and dirty is occasionally the right call.

> Also that is very much a "I don't need seatbelts; I'm a good driver" attitude.

Lol, fair point... Again, I'm in favor of this as an optional rule and would use it in some projects.

> Then they are misguided. Many (if not most) Python devs consider static typing to be a waste of time entirely.

I _think_ that's changing? But that could be because I adopted it years ago and now take it for granted. And I agree the `if id:` is not only misguided but just plain wrong in any use case where a future `int` conversion is an obvious possibility (e.g. when the original `str` value is already a numeric).
